File: examples/go/02_transactions.go
// Transaction Example
//
// Demonstrates ACID transactions:
// - Using WithTransaction for automatic commit/rollback
// - Read operations within transactions
// - Error handling and rollback

package main

import (
	"fmt"
	"log"
	"strconv"

	toondb "github.com/toondb/toondb-go"
)

func main() {
	db, err := toondb.Open("./txn_example_db")
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()
	fmt.Println("✓ Database opened")

	// Automatic transaction with callback (recommended)
	err = db.WithTransaction(func(txn *toondb.Transaction) error {
		if err := txn.Put("accounts/alice/balance", []byte("1000")); err != nil {
			return err
		}
		if err := txn.Put("accounts/bob/balance", []byte("500")); err != nil {
			return err
		}
		fmt.Println("✓ Transaction: wrote initial balances")
		return nil
	})
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println("✓ Transaction committed automatically")

	// Simulate a transfer
	err = db.WithTransaction(func(txn *toondb.Transaction) error {
		// Read current balances
		aliceData, _ := txn.Get("accounts/alice/balance")
		bobData, _ := txn.Get("accounts/bob/balance")

		aliceBalance, _ := strconv.Atoi(string(aliceData))
		bobBalance, _ := strconv.Atoi(string(bobData))

		transferAmount := 250

		// Update balances atomically
		if err := txn.Put("accounts/alice/balance", 
			[]byte(strconv.Itoa(aliceBalance-transferAmount))); err != nil {
			return err
		}
		if err := txn.Put("accounts/bob/balance", 
			[]byte(strconv.Itoa(bobBalance+transferAmount))); err != nil {
			return err
		}

		fmt.Printf("✓ Transfer: Alice -> Bob: $%d\n", transferAmount)
		return nil
	})
	if err != nil {
		log.Fatal(err)
	}

	// Verify final balances
	alice, _ := db.Get("accounts/alice/balance")
	bob, _ := db.Get("accounts/bob/balance")

	fmt.Println("\n✓ Final balances:")
	fmt.Printf("  Alice: $%s\n", string(alice))
	fmt.Printf("  Bob: $%s\n", string(bob))

	// Transaction rollback on error
	err = db.WithTransaction(func(txn *toondb.Transaction) error {
		txn.Put("accounts/alice/balance", []byte("9999"))
		return fmt.Errorf("simulated failure")
	})
	if err != nil {
		fmt.Printf("✓ Transaction rolled back: %v\n", err)
	}

	// Verify balance unchanged after rollback
	aliceAfter, _ := db.Get("accounts/alice/balance")
	fmt.Printf("✓ Alice's balance after rollback: $%s\n", string(aliceAfter))
}

File: examples/rust/02_transactions.rs
//! Transaction Example
//! 
//! This example demonstrates ACID transactions:
//! - Using with_transaction for automatic commit/rollback
//! - Manual transaction control
//! - Read operations within transactions

use sochdb::Database;
use anyhow::Result;

fn main() -> Result<()> {
    let db = Database::open("./txn_example_db")?;
    println!("✓ Database opened");

    // Automatic transaction with closure (recommended)
    db.with_transaction(|txn| {
        txn.put(b"accounts/alice/balance", b"1000")?;
        txn.put(b"accounts/bob/balance", b"500")?;
        println!("✓ Transaction: wrote initial balances");
        Ok(())
    })?;
    println!("✓ Transaction committed automatically");

    // Simulate a transfer
    db.with_transaction(|txn| {
        // Read current balances
        let alice_balance: i64 = txn.get(b"accounts/alice/balance")?
            .map(|v| String::from_utf8_lossy(&v).parse().unwrap_or(0))
            .unwrap_or(0);
        let bob_balance: i64 = txn.get(b"accounts/bob/balance")?
            .map(|v| String::from_utf8_lossy(&v).parse().unwrap_or(0))
            .unwrap_or(0);

        let transfer_amount = 250;

        // Update balances
        txn.put(b"accounts/alice/balance", 
                (alice_balance - transfer_amount).to_string().as_bytes())?;
        txn.put(b"accounts/bob/balance", 
                (bob_balance + transfer_amount).to_string().as_bytes())?;

        println!("✓ Transfer: Alice -> Bob: ${}", transfer_amount);
        Ok(())
    })?;

    // Verify final balances
    let alice = db.get(b"accounts/alice/balance")?
        .map(|v| String::from_utf8_lossy(&v).to_string())
        .unwrap_or_default();
    let bob = db.get(b"accounts/bob/balance")?
        .map(|v| String::from_utf8_lossy(&v).to_string())
        .unwrap_or_default();

    println!("\n✓ Final balances:");
    println!("  Alice: ${}", alice);
    println!("  Bob: ${}", bob);

    // Transaction rollback on error
    let result = db.with_transaction(|txn| {
        txn.put(b"accounts/alice/balance", b"9999")?;
        // Simulate an error
        Err(anyhow::anyhow!("Simulated failure"))
    });

    match result {
        Ok(_) => println!("Transaction committed"),
        Err(e) => println!("✓ Transaction rolled back: {}", e),
    }

    // Verify balance unchanged after rollback
    let alice_after = db.get(b"accounts/alice/balance")?
        .map(|v| String::from_utf8_lossy(&v).to_string())
        .unwrap_or_default();
    println!("✓ Alice's balance after rollback: ${}", alice_after);

    Ok(())
}
